sdks/java/extensions/smb/src/main/java/org/apache/beam/sdk/extensions/smb/BucketSourceIterator.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.smb;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.Iterator;
+import org.apache.beam.sdk.coders.AtomicCoder;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.extensions.smb.SortedBucketFile.Reader;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.io.fs.ResourceIdCoder;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.TupleTag;
+
+/**
+ * Iterates over shards in a bucket one record at a time.
+ *
+ * @param <KeyT>
+ */
+class BucketSourceIterator<KeyT> implements Serializable {
+  private final TupleTag tupleTag;
+  private final Reader<?> reader;
+  private final ResourceId resourceId;
+  private final BucketMetadata<KeyT, Object> metadata;
+
+  private KV<KeyT, ?> nextKv;
+
+  BucketSourceIterator(
```

Comment:
We can probably do the key type unificiation here.

- change `BucketMetadata#extractSortingKey` to return `Array[Byte]`
- store key bytes in `nextKv` and compare bytes in the iterator
- add a `keyClass` param and infer coder
- decode key bytes to `keyClass` at the end of `nextKeyGroup`

Reply:
Actually we can move the unification up to `SortedBucketSource`, and return `Array[Byte]` key here.